File: exercism/python/sieve/sieve.py
from itertools import islice
import math


def sieve(upto):
    primes = [None, None] + list(range(2, upto + 1))
    p = 2
    while True:
        primes[2*p::p] = [None] * math.floor((upto - p) / p)
        try:
            p = next(filter(identity, islice(primes, p + 1, None)))
        except StopIteration:
            return list(filter(identity, primes))


identity = lambda x: x

File: exercism/csharp/sieve/Sieve.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Sieve
{
    public static IEnumerable<int> Primes (int upto)
    {
        var candidates = Enumerable.Range(0, upto + 1).ToArray();
        var p = 2;
        yield return p;
        while (p < upto) {
            // null out divisors
            for (var q = 2; q < (double)upto / p; q++) {
                candidates[p * q] = 0;
            }
            // increase p to next known prime
            if (p > 2) yield return p;
            p++;
            while (candidates[p] == 0) p++;
        }
    }
}
